datalad/support/gitrepo.py

```diff
@@ -3644,16 +3668,24 @@ def _get_cache_key(label, paths, ref, untracked=None):
                         # report the shasum that we know, for further
                         # wrangling of subdatasets below
                         props['gitshasum'] = to_state_r['gitshasum']
-                        props['prev_gitshasum'] = from_state[f]['gitshasum']
+                        props['prev_gitshasum'] = fr_state['gitshasum']
             else:
                 # change in git record, or on disk
                 props = dict(
                     # TODO we could have a new file that is already staged
                     # but had subsequent modifications done to it that are
                     # unstaged. Such file would presently show up as 'added'
                     # ATM I think this is OK, but worth stating...
-                    state='modified' if f.exists() or \
-                    f.is_symlink() else 'deleted',
+                    # re `None`: if we face a dataset that git reports to
+                    # have a changed state, we cannot rely on this judgement
+                    # with subdatasets that are potentially in adjusted mode.
+                    # Setting the `state` to None will queue this record for
+                    # further processing below.
```

Comment:
In general I find this and related functions hard to grok, so your comments are very much appreciated.